File: tiq-e/trees/2/main.py
from time import perf_counter


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isValidBST(self, root,  lessThan=float('inf'), largerThan=float('-inf')):
        if root is None:
            return True

        elif root.val >= lessThan or root.val <= largerThan:
            return False

        return self.isValidBST(root.left, min(lessThan, root.val), largerThan) and self.isValidBST(root.right, lessThan,
                                                                                                   max(largerThan,
                                                                                                       root.val))


root = TreeNode(5, None, None)

start = perf_counter()
result = Solution().isValidBST(root)
end = perf_counter()
print("Duration: " + str(end - start))
print("Result: " + str(result))

File: tiq-e/trees/2/main.ts
import { performance } from "perf_hooks";

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;

  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function isValidBST(
  root: TreeNode,
  min: TreeNode | null = null,
  max: TreeNode | null = null
): boolean {
  if (!root) return true;

  if (min && root.val <= min!.val) return false;
  if (max && root.val >= max!.val) return false;

  return (
    isValidBST(root.left as TreeNode, min, root) &&
    isValidBST(root.right as TreeNode, root, max)
  );
}

const root = new TreeNode(5, null, null);

const start = performance.now();
const result = isValidBST(root);
const end = performance.now();
console.log("Duration:", end - start);
console.log("Result:", result);
